src/object_store/presign.nim

```diff
@@ -0,0 +1,176 @@
+  trace("object store: " & $httpMethod & " " & $url & " -> " & $signResponse.code)
+
+  if signResponse.code notin [Http302, Http307]:
+    trace("object store: " & signResponse.body())
+    raise newException(ValueError, "Presign requires 302/307 redirect but received: " & signResponse.status)
+
+  if not signResponse.headers.hasKey("location"):
+    raise newException(ValueError, "Presign edirect Location header missing")
+
+  let uri = parseUri(signResponse.headers["location"])
+  if uri.scheme == "":
+    trace("object store: " & $uri)
+    raise newException(ValueError, "Presign edirect Location header needs to be absolute URL")
+
+  var
+    names   = newSeq[string]()
+    headers = newHttpHeaders(@[
+      ("Content-Type", "application/json"),
+      ("Content-Length", $len(body)),
+    ])
+  if signResponse.headers.hasKey("x-forward-headers"):
+    names = signResponse.headers["x-forward-headers"].strip().split(',')
+    for i in names:
+      let name = i.strip()
+      if signResponse.headers.hasKey(name):
+        headers[name] = signResponse.headers[name]
+
+  trace("object store: " & $httpMethod & " @" & uri.hostname & " (" & $len(body) & "bytes) forwarding: " & $names)
+  let response = safeRequest(url               = uri,
+                             headers           = headers,
+                             timeout           = self.timeout,
+                             disallowHttp      = self.disallowHttp,
+                             pinnedCert        = self.pinnedCertFile,
+                             httpMethod        = httpMethod,
+                             body              = body,
+                             retries           = 2,
+                             firstRetryDelayMs = 100,
+                             raiseWhenAbove    = 500)
+  trace("object store: " & $httpMethod & " @" & $uri.hostname & " (" & $len(body) & "bytes) -> " & $response.code)
+
+  let updatedRef = deepCopy(keyRef)
+  updatedRef.query = signResponse.headers.getOrDefault("x-chalk-object-query")
+  return (response, updatedRef)
+
+proc objectExists(this: ObjectStoreConfig, keyRef: ObjectStoreRef): ObjectStoreRef =
+  let
+    self = ObjectStorePresign(this)
+    (response, updatedRef) = self.request(
+      keyRef,
+      auth       = self.readAuth,
+      httpMethod = HttpHead,
+      body       = "",
+    )
+  if response.code.is2xx():
+    return updatedRef
+  elif response.code == Http404:
+    return nil
+  else:
+    raise newException(ValueError, $response.code & " " & response.body())
+
+proc createObject(this: ObjectStoreConfig, keyRef: ObjectStoreRef, data: string): ObjectStoreRef =
+  let
+    self = ObjectStorePresign(this)
+    (response, updatedRef) = self.request(
+      keyRef,
+      auth       = self.writeAuth,
+      httpMethod = HttpPut,
+      body       = data,
+    )
+  if response.code.is2xx():
+    return updatedRef
+  else:
+    raise newException(ValueError, $response.code & " " & response.body())
```

Comment:
Object store API URLs can be finicky when initially configuring. Additional logging to handle 4xx errors could improve user experience.